synapseclient/services/json_schema.py

```diff
@@ -0,0 +1,613 @@
+        resource_access = [
+            {"principalId": principal_id, "accessType": access_type}
+            for principal_id in principal_ids
+        ]
+        response = self.service.update_organization_acl(self.id, resource_access, etag)
+        return response
+
+    def update_acl(
+        self,
+        principal_ids: Sequence[int],
+        access_type: Sequence[str] = DEFAULT_ACCESS,
+        etag: str = None,
+    ):
+        self.must_get()
+        principal_ids = set(principal_ids)
+        acl = self.get_acl()
+        resource_access = acl["resourceAccess"]
+        if etag is None:
+            etag = acl["etag"]
+        for entry in resource_access:
+            if entry["principalId"] in principal_ids:
+                entry["accessType"] = access_type
+                principal_ids.pop(entry["principalId"])
+        for principal_id in principal_ids:
+            entry = {
+                "principalId": principal_id,
+                "accessType": access_type,
+            }
+            resource_access.append(entry)
+        response = self.service.update_organization_acl(self.id, resource_access, etag)
+        return response
+
+    def list_json_schemas(self):
+        self.must_get()
+        response = self.service.list_json_schemas(self.name)
+        for raw_json_schema in response:
+            json_schema = JsonSchema.from_response(self, raw_json_schema)
+            self._raw_json_schemas[json_schema.name] = raw_json_schema
+            self._json_schemas[json_schema.name] = json_schema
+            yield json_schema
+
+    def get_json_schema(self, json_schema_name, raw: bool = False):
+        self.must_get()
+        if json_schema_name not in self._json_schemas:
+            list(self.list_json_schemas())
+        if raw:
+            json_schema = self._raw_json_schemas.get(json_schema_name)
+        else:
+            json_schema = self._json_schemas.get(json_schema_name)
+        return json_schema
+
+    def create_json_schema(
+        self,
+        json_schema_body,
+        name: str = None,
+        semantic_version: str = None,
+        dry_run: bool = False,
+    ):
+        if name:
+            uri = f"{self.name}-{name}"
+            if semantic_version:
+                uri = f"{uri}-{semantic_version}"
+            json_schema_body["$id"] = uri
+        else:
+            assert (
+                semantic_version is not None
+            ), "Specify both the name and the semantic version (not just the latter)"
+        response = self.service.create_json_schema(json_schema_body, dry_run)
+        if dry_run:
+            return response
+        raw_version = response["newVersionInfo"]
+        json_schema = JsonSchemaVersion.from_response(self, raw_version)
+        return json_schema
+
+
+class JsonSchemaService:
+    def __init__(self, synapse: Synapse = None) -> None:
+        self.synapse = synapse
+
+    @wraps(Synapse.login)
+    def login(self, *args, **kwargs):
```

Comment:
Hmmm - what is your rationale for providing the option for each service to have its own login function, but also allowing users to specify their own synapse object?

Reply:
Because users can theoretically create a `JsonSchemaService` instance without running `synapse.service("json_schema")`, I wanted to make sure there was a method to "log in" a `JsonSchemaService` instance after the fact. Previously, I had a `JsonSchemaService.set_synapse()` method, but I find this login method more elegant.